File: CostVisitor.cpp
#include "CostVisitor.h"
#include "Tool.h"
#include "Grocery.h"

void CostVisitor::visit(Hammer& hammer) {
    double priceWithTax = hammer.GetCost() * (1.0 + GST);
    totalCost += priceWithTax;
    toolCost += priceWithTax;
}

void CostVisitor::visit(Screwdriver& screwdriver) {
    double priceWithTax = screwdriver.GetCost() * (1.0 + GST);
    totalCost += priceWithTax;
    toolCost += priceWithTax;
}

void CostVisitor::visit(Apple& apple) {
    totalCost += apple.GetCost();
    groceryCost += apple.GetCost();
}

void CostVisitor::visit(Orange& orange) {
    totalCost += orange.GetCost();
    groceryCost += orange.GetCost();
}

double CostVisitor::getTotalCost() const { return totalCost; }
double CostVisitor::getToolCost() const { return toolCost; }
double CostVisitor::getGroceryCost() const { return groceryCost; }

void CostVisitor::reset() {
    totalCost = 0.0;
    toolCost = 0.0;
    groceryCost = 0.0;
}

File: CostVisitor.java
public class CostVisitor implements Visitor {
    private double totalCost = 0.0;
    private double toolCost = 0.0;
    private double groceryCost = 0.0;
    public static final double GST = 0.14;

    public void visit(Hammer hammer) {
        double priceWithTax = hammer.getCost() * (1.0 + GST);
        totalCost += priceWithTax;
        toolCost += priceWithTax;
    }

    public void visit(Screwdriver screwdriver) {
        double priceWithTax = screwdriver.getCost() * (1.0 + GST);
        totalCost += priceWithTax;
        toolCost += priceWithTax;
    }

    public void visit(Apple apple) {
        totalCost += apple.getCost();
        groceryCost += apple.getCost();
    }

    public void visit(Orange orange) {
        totalCost += orange.getCost();
        groceryCost += orange.getCost();
    }

    public double getTotalCost() { return totalCost; }
    public double getToolCost() { return toolCost; }
    public double getGroceryCost() { return groceryCost; }
    public void reset() {
        totalCost = 0.0;
        toolCost = 0.0;
        groceryCost = 0.0;
    }
}
